[SPARK-15346] [MLlib] Reduce duplicate computation in picking initial points

mouendless opened the pull request:
@mateiz @srowen 

I state that the contribution is my original work and that I license the work to the project under the project's open source license

There's some format problems with my last PR, with @HyukjinKwon 's help I read the guidance, re-check my code and PR, then run the tests, finally re-submit the PR request here.

The related JIRA issue though marked as resolved, this change may relate to it I think.
## Proposed Change

After picking each new initial centers, it's unnecessary to compute the distances between all the points and the old ones. 
Instead this change keeps the distance between all the points and their closest centers, and compare to the distance of them with the new center then update them.
## Test result

One can find an easy test way in (https://issues.apache.org/jira/browse/SPARK-6706)

I test the KMeans++ method for a small dataset with 16k points, and the whole KMeans|| with a large one with 240k points.
The data has 4096 features and I tunes K from 100 to 500.
The test environment was on my 4 machine cluster, I also tested a 3M points data on a larger cluster with 25 machines and got similar results, which I would not draw the detail curve. The result of the first two exps are shown below
### Local KMeans++ test:

Dataset:4m_ini_center
Data_size:16234
Dimension:4096 

Lloyd's Iteration = 10
The y-axis is time in sec, the x-axis is tuning the K.

![image](https://cloud.githubusercontent.com/assets/10915169/15175831/d0c92b82-179a-11e6-8b68-4e165fc2fdff.png)

![local_total](https://cloud.githubusercontent.com/assets/10915169/15175957/6b21c3b0-179b-11e6-9741-66dfe4e23eb7.jpg)
### On a larger dataset

An improve show in the graph but not commit in this file: In this experiment I also have an improvement for calculation in normalization data (the distance is convert to the cosine distance). As if the data is normalized into (0,1), one improvement in the original vesion for util.MLUtils.fastSauaredDistance would have no effect (the precisionBound 2.0 \* EPSILON \* sumSquaredNorm / (normDiff \* normDiff + EPSILON) will never less then precision in this case). Therefore I design an early terminal method when comparing two distance (used for findClosest). But I don't include this improve in this file, you may only refer to the curves without "normalize" for comparing the results.

Dataset:4k24  
Data_size:243960
Dimension:4096

Normlize    Enlarge     Initialize  Lloyd's_Iteration
NO      1            3            5
YES             10000    3            5

Notice: the normlized data is enlarged to ensure precision

The cost time: x-for value of K, y-for time in sec
![4k24_total](https://cloud.githubusercontent.com/assets/10915169/15176635/9a54c0bc-179e-11e6-81c5-238e0c54bce2.jpg)

SE for unnormalized data between two version, to ensure the correctness
![4k24_unnorm_se](https://cloud.githubusercontent.com/assets/10915169/15176661/b85dabc8-179e-11e6-9269-fe7d2101dd48.jpg)

Here is the SE between normalized data just for reference, it's also correct.
![4k24_norm_se](https://cloud.githubusercontent.com/assets/10915169/15176742/1fbde940-179f-11e6-8290-d24b0dd4a4f7.jpg)

mouendless committed:
Reduce duplicate computation

Reduce duplicate computation when picking initial points

AmplabJenkins commented:
Can one of the admins verify this patch?

srowen commented:
@mouendless this does not appear to be directly related to SPARK-6706, and it has a lot of style problems.

After a first pass, I don't think this is valid. You have changed the computation to ignore the point's distance to previous centers. Right? Of course that's faster, but I don't believe it's correct.

mouendless commented:
@srowen Thanks for taking time reviewing and sry for style problem, I will re-check for it.

But logically, I don't ignore the point distance to previous centers, 
KMeans.pointCost(curCenters, p) computes the shortest distance between p towards all the current centers. As during the initial center picking phases, the centers do not move but new ones are added. We can keep in memory for every point the closest distance towards its closest center, and each step compare it with the distance towards new center. If the new one is closer, update it. 

I compare the result of sum and cumulativeScore computed by the two version, they are totally the same, and that are the only things may affect the reuslt correctness in the new version. I also compared the SE computed from this two version, they are still the same. Therefore the same result with less time in experiment in my opinion

srowen commented:
OK I see what you're saying, yes. This could be fine. Make a new JIRA though. I will add some comments.

srowen commented:
Jenkins test this please

srowen commented on mllib/src/main/scala/org/apache/spark/mllib/clustering/LocalKMeans.scala:
```diff
@@ -46,17 +46,20 @@ private[mllib] object LocalKMeans extends Logging {
 
     // Initialize centers by sampling using the k-means++ procedure.
     centers(0) = pickWeighted(rand, points, weights).toDense
+    val costArray = points.map { point =>
```
can be `.map(KMeans.fastSquaredDistance(_, centers(0)))` ? I forget whether this is actually allowed as syntax

srowen commented on mllib/src/main/scala/org/apache/spark/mllib/clustering/LocalKMeans.scala:
```diff
@@ -46,17 +46,20 @@ private[mllib] object LocalKMeans extends Logging {
-      val curCenters = centers.view.take(i)
-      val sum = points.view.zip(weights).map { case (p, w) =>
-        w * KMeans.pointCost(curCenters, p)
+      val sum = costArray.view.zip(weights).map{
```
Can be one line, I think, and needs space before brace. Maybe as simple to just `.map(p => p._1 * p._2)` but a case is OK too. I don't think a `.view` on a local array helps.

srowen commented on mllib/src/main/scala/org/apache/spark/mllib/clustering/LocalKMeans.scala:
```diff
@@ -66,6 +69,12 @@ private[mllib] object LocalKMeans extends Logging {
       }
+
+      //update costArray
+      for(p <- 0 to points.length-1){
```
`for (p <- 0 until points.length)` or better `for (p <- points.indices)`

srowen commented on mllib/src/main/scala/org/apache/spark/mllib/clustering/LocalKMeans.scala:
```diff
@@ -66,6 +69,12 @@ private[mllib] object LocalKMeans extends Logging {
+
+      //update costArray
+      for(p <- 0 to points.length-1){
+        costArray(p)=math.min(KMeans.fastSquaredDistance(points(p), centers(i)),costArray(p))
```
missing spaces around operations, args

mouendless replied:
JIRA already created and linked, 
I will fix this once I have access to my work machine

SparkQA commented:
**[Test build #58640 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58640/consoleFull)** for PR 13133 at commit [`c28db67`](https://github.com/apache/spark/commit/c28db67f7bd3257ac87580072228fd9854a3c883).

SparkQA commented:
**[Test build #58640 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58640/consoleFull)** for PR 13133 at commit [`c28db67`](https://github.com/apache/spark/commit/c28db67f7bd3257ac87580072228fd9854a3c883).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/58640/
Test FAILed.

HyukjinKwon commented:
(Please let me leave a link I usually refer, https://github.com/databricks/scala-style-guide)

mateiz commented:
Cool, the change does look right to me, but as Sean said there are some style issues. It should definitely help speed up initialization!

mouendless committed:
some code style fix

There is still some problem with my style check plugin, still need to be further check

mouendless committed:
scala style fix

mouendless commented:
@srowen progress report. I finally fix my scala style plugin. I ran it and fixed the reported style issues ,and compile, please check again. Though the code can be submit to my cluster and complete the corresponding test on my dataset, I will run again the scalatest:test.

srowen commented on mllib/src/main/scala/org/apache/spark/mllib/clustering/LocalKMeans.scala:
```diff
@@ -46,17 +46,15 @@ private[mllib] object LocalKMeans extends Logging {
-      val sum = points.view.zip(weights).map { case (p, w) =>
-        w * KMeans.pointCost(curCenters, p)
-      }.sum
+      val sum = costArray.view.zip(weights).map(p => p._1 * p._2).sum
```
Remove .view?

mouendless replied:
Oh yes, it's now only an int array and unnecessary with view

srowen commented on mllib/src/main/scala/org/apache/spark/mllib/clustering/LocalKMeans.scala:
```diff
@@ -66,6 +64,12 @@ private[mllib] object LocalKMeans extends Logging {
       }
+
+      // update costArray
+      for(p <- points.indices) {
```
space after for

mouendless committed:
style issue

mouendless committed:
remove unnecessary .view for costArray

costArray.view  -> costArray

srowen commented:
Jenkins retest this please

SparkQA commented:
**[Test build #58768 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58768/consoleFull)** for PR 13133 at commit [`3a0f4e6`](https://github.com/apache/spark/commit/3a0f4e6f54a00d785b3c4626347e6a0b74608c26).

SparkQA commented:
**[Test build #58768 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58768/consoleFull)** for PR 13133 at commit [`3a0f4e6`](https://github.com/apache/spark/commit/3a0f4e6f54a00d785b3c4626347e6a0b74608c26).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/58768/
Test PASSed.

srowen commented:
Merged to master/2.0
